### PR TITLE
Structure_Engine: Add engine methods for retaining wall classes

### PR DESCRIPTION
**Pull request by peterjamesnugent**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/1621
   
### Issues addressed by this PR

Closes [BHoM_Structure_oM](https://github.com/BHoM/BHoM/issues/1619#issue-2388764920)

### Test files

See. 
https://github.com/BHoM/BHoM/pull/1621

### Changelog

Added required extension methods for implemented interfaces for new retaining wall classes.
- Added `Geometry()` method for `Stem` and `RetainingWall`
- Added `IsNull()` method for `Stem` and `RetainingWall`
- Added `SolidVolume()` method for `Stem` and `RetainingWall`
- Added `MaterialComposition()` method for `Stem` and `RetainingWall`
- Added `SetOutlineElements(List<IElement1D> outlineElements1D)` method for `Stem`
- Added `OutlineElemets1D()` method for `Stem`
- Added `Geometry3D()` method for `Stem` and `RetainingWall`
- Migrated `IsQuad`, `IsTriangular`, `IsRectangular`, `IsSquare` to the `Geometry` namespace so that it can be used on `PolyCurve`s - the methods in `BH.Engine.Analytical` are unchanged, and just reference the `Geometry` methods.

### Additional comments

`HasMargeablePropertiesWith(IElement2D)` has not been added for any of the objects in reference to:
https://github.com/BHoM/BHoM_Engine/pull/3138#discussion_r1312778359

### CODE REVIEW

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check ready-to-merge

**Comment by peterjamesnugent** [comment]
Discussed with @IsakNaslundBh. I have reviewed the PR raised by @KalleEdstroem including testing using the test-scripts.

My commits relate to updating the methods because the changes to PadFoundation and Stem making use of IRegion and the versioning required.

File: Structure_Engine/Query/SolidVolume.cs [hunk 1]
```diff
@@ -106,6 +104,26 @@ public static double SolidVolume(this PileFoundation pileFoundation)
...
         }
 
+        /***************************************************/
+
+        [Description("Returns a Stem's solid volume based on the area of the element and its top and bottom thickness.")]
+        [Input("stem", "The Stem to get the volume from.")]
+        [Output("volume", "The Stem solid material volume.", typeof(Volume))]
+        public static double SolidVolume(this Stem stem)
+        {
+            return stem.IsNull() ? 0 : Engine.Geometry.Query.IArea(stem.Perimeter) * (stem.ThicknessBottom + stem.ThicknessTop) / 2;
```
Comment (IsakNaslundBh) [inline Structure_Engine/Query/SolidVolume.cs:114]: Think this is only true for a rectangular perimiter (with some other requirements on it as well) as well as some paralellogram shapes etc. For, for example, a perimiter looking something like: 
image
The volume would be incorrect.

Could see this fixed in a couple of different ways:
1. Fully correctly compute the volume for any perimiter shape (dont think this is feasible in this PR)
2. Change the class to enforce rectangular outlines, and then state that the bottom/top thickness relates to that rectangles coordinate system. (might be the msot involved change, and also a bit less flexible as I presume there is a reason you allowed for a PolyCurve/ICurve in the first place)
3. Add checks to see if the formula is correct:
   - Is top and bottom thickness equal, if so, the formula is correct
   - Is the perimiter rectangular and aligned with what is assumed to be top and bottom gradient? If so, formula is correct
   - Else, either raise a warning, stating something about that the volume is approximated, and should not be fully trusted, or raise an error and return NaN. 
4. Always raise a warning stating the requirements, and warn about that it sometimes can be invalid. (quickest to do, but also worst UX I think)

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            if (!line.IsInPlane(Geometry.Create.Plane(line.Start, Vector.ZAxis)))
+            {
+                Base.Compute.RecordError("Provided line is not parallel to the XY plane. Please provide a line parallel to the XY plane.");
+                return null;
+            }
+
+            if (!line.IsInPlane(Geometry.Create.Plane(line.Start, Vector.ZAxis)))
+            {
+                Base.Compute.RecordError("Provided line is not parallel to the XY plane. Please provide a line parallel to the XY plane.");
+                return null;
```
Comment (IsakNaslundBh) [inline Structure_Engine/Create/Elements/RetainingWall.cs:98]:
```suggestion
```
Same check is provided twice? Should remove one

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            }
+
+            PolyCurve stemOutline = new PolyCurve();
+            PolyCurve footingOutline = new PolyCurve();
+
+            //Create the footing outline. 
+            Line toeLine = line.ShallowClone();
+            Line heelLine = line.ShallowClone();
```
Comment (IsakNaslundBh) [inline Structure_Engine/Create/Elements/RetainingWall.cs:106]: Guess this does not really hurt, other than a minor, most likely unimportant slowdown, but do not need these clones. Translate returns a new line, and does not modify the incoming line.

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+            toeLine = toeLine.Translate(normal * (toeLength + stemThickness / 2));
+            heelLine = heelLine.Translate(normal * (heelLength + stemThickness / 2) * -1).Reverse();
+
+            footingOutline.Curves = new List<ICurve> { toeLine, Geometry.Create.Line(toeLine.End, heelLine.Start), heelLine, Geometry.Create.Line(heelLine.End, toeLine.Start) };
+
+            //Create the Stem outline.
+            Line bottomLine = line.ShallowClone();
+            Line topLine = line.ShallowClone();
```
Comment (IsakNaslundBh) [inline Structure_Engine/Create/Elements/RetainingWall.cs:115]: As above, don't think you need to pre-clone like this.

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Some comments below. Mostly the Volume calculation that is a bit iffy. The Create method comments are not as bad, even if the double check for planarity most likely needs to go.

File: Structure_Engine/Query/SolidVolume.cs (hunk 1 above)
Comment (peterjamesnugent) [reply Structure_Engine/Query/SolidVolume.cs:114]: I have opted for Option 3 and raised an issue here for future development: https://github.com/BHoM/BHoM_Engine/issues/3409

**Comment by peterjamesnugent** [comment]
@BHoMBot check versioning
@BHoMBot check unit-tests

**Comment by pawelbaran** [comment]
@BHoMBot check versioning
@BHoMBot check unit-tests

**Comment by pawelbaran** [comment]
@BHoMBot check versioning
@BHoMBot check unit-tests

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check required

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests

**Review by IsakNaslundBh — APPROVED** [review]
Happy with the changes made and hence happy to approve this PR.

**Pull request merged**